Module-5/Task List/Task List/Model/Store.swift

```diff
@@ -0,0 +1,29 @@
+//
+//  Store.swift
+//  Created by Mark Renaud (2023).
+//
+
+import Foundation
+
+class Store: ObservableObject {
+    /// All tasks saved to the list.
+    @Published var tasks: [TaskItem]
+
+    init(tasks: [TaskItem] = []) {
+        self.tasks = tasks
+    }
+
+    func addTask(_ task: TaskItem) {
+        tasks.append(task)
+    }
+    
+    func complete(_ taskID: UUID) {
+        if let storeIndex = tasks.firstIndex(where: { $0.id == taskID }) {
+            tasks[storeIndex].isCompleted = true
+        }
+    }
+    
+    func complete(_ task: TaskItem) {
+        complete(task.id)
+    }
```

Comment:
So we have some unused methods here? ;)

Reply:
I probably should explain what I've done here.

I included the `complete(_:)` method as it was a requirement in the assignment notes.

I chose not to use this method because I believe it represents an anti-pattern in SwiftUI.

Here is my logic:

- I believe the logic for including such a method in the `Store` is a "leftover" from UIKit + MVVM days.  The aim would be that by routing the marking of completion through such a method, there is a single place for implementing associated business logic - and that logic can be unit tested.
- As the app currently stands, no additional logic occurs when a `TaskItem` is marked as completed. Requiring the `EditView` to update via the `Store` forces the view to unnecessarily know about the `Store` (what happens if the `EditView` wanted to edit a `TaskItem` that is not yet in the store?).  It also requires the `EditView` to implement unnecessary logic binding to the Store and TaskItem to the toggle.  For example, the Store and EditView would need to be implemented like:

```Swift
import SwiftUI

class Store: ObservableObject {
    /// All tasks saved to the list.
    @Published var tasks: [TaskItem]

    init(tasks: [TaskItem] = []) {
        self.tasks = tasks
    }

    func addTask(_ task: TaskItem) {
        tasks.append(task)
    }
    
    func mark(task: TaskItem, completed: Bool) {
        if let taskIndex = tasks.firstIndex(where: { $0.id == task.id}) {
            self.tasks[taskIndex].isCompleted = completed
        }
    }
}

struct EditTaskView: View {
    @EnvironmentObject var store: Store
    @State private var toggleState: Bool = false
    @Binding var task: TaskItem

    var body: some View {
        Form {
            Section("Task Title") {
                TextField("Task Title", text: $task.title)
            }
            Section("Notes") {
                TextField("Notes", text: $task.notes, axis: .vertical)
                    .lineLimit(5 ... 10)
            }
            Section {
                Toggle("Completed:", isOn: $toggleState)
                    .onAppear {
                        toggleState = task.isCompleted
                    }
                    .onChange(of: toggleState) { newValue in
                        store.mark(task: task, completed: newValue)
                    }
            }
        }
    }
}

```

- We should consider the simplest way for SwiftUI to manipulate the data model without knowing unnecessary details, working against the built-in bindings, and while increasing the reusability of the view.
- If the `Store` needs to perform some custom logic on a change in one of the `TaskItem`s **it owns**, I believe it should observe itself for such changes and then act accordingly.  This approach maintains boundaries, allows for clean SwiftUI bindings, and retains testability.  I've added a commit to this branch, which includes a unit test, for your review. Here is the essence of it:

```Swift

// -- View doesn't need to care where the task comes from (or if it is in a Store) and
//    can bind directly to the TaskItem model.

import SwiftUI

struct EditTaskView: View {
    @Binding var task: TaskItem
    
    var body: some View {
        Form {
            Section("Task Title") {
                TextField("Task Title", text: $task.title)
            }
            Section("Notes") {
                TextField("Notes", text: $task.notes, axis: .vertical)
                    .lineLimit(5...10)
            }
            Section {
                Toggle("Completed:", isOn: $task.isCompleted)
            }
        }
    }
}

// -- tore observes for changes of any tasks it owns and adds custom business
//    logic when observed.

import Foundation

class Store: ObservableObject {
    /// All tasks saved to the list.
    @Published var tasks: [TaskItem] { 
        didSet {
            implementCustomChangeLogic()
        }
    }

    init(tasks: [TaskItem] = []) {
        self.tasks = tasks
    }

    func addTask(_ task: TaskItem) {
        tasks.append(task)
    }
    
//  The following is not part of the assignment.  It is in relation
//  to discussion re implementation details in assignment review
//  pull request.

    private (set) var taskChangesSinceLaunch: Int = 0
    
    private func implementCustomChangeLogic() {
        // it is the Store's responsibility to monitor for change
        // to itself
        print("Custom logic on task change here")
        taskChangesSinceLaunch += 1
    }
}

--- Business logic remains testable and separate from the View
@testable import Task_List
import XCTest

final class Task_ListTests: XCTestCase {
    var sampleStore = Store(tasks: .sampleTasks)

    func testTasksChanged() throws {
        let preChangeCount = sampleStore.taskChangesSinceLaunch
        
        // update completion state of first task
        sampleStore.tasks[0].isCompleted.toggle()
        
        let postChangeCount = sampleStore.taskChangesSinceLaunch
        let expectedPostChangeCount = preChangeCount + 1
        
        XCTAssertEqual(expectedPostChangeCount, postChangeCount)
    }
}
``` 

I am eager to hear your thoughts on this logic and your view on how this problem can best be approached before I merge back into main. 
It is issues like this where I would appreciate having colleagues with whom to bounce ideas and patterns back and forth.